### ISSUE TITLE
Update user's documentation 

### ISSUE THREAD

**Issue by AndyNicks** [ably/ably-asset-tracking-android#630]
Review and update https://ably.com/documentation/asset-tracking to reflect all of the updates we made during last sprints

┆Issue is synchronized with this [Jira Task](https://ably.atlassian.net/browse/SDK-1577) by [Unito](https://www.unito.io)

**Comment by KacperKluka** [comment]
I read our current docs and based on that and our current API I created a list of improvements for the docs.

-----

https://ably.com/docs/asset-tracking

**2. Resolution**
- Maybe we should write, that even though the `desiredInterval` does not work for GPS resolution on iOS, it still works and affects the network resolution (how often location updates are sent to the subscribers)?

**3. Resolution set**
- I'd rephrase the first sentence, so it is pointed out that the set is optional for the resolution policy to work.
> A resolution set, is a set of resolutions that, when provided, can be used by the resolution policy to create the best possible resolution for a given situation.

**3.1 Default resolution set**
- I'd point out that while the set is required to create the default resolution constraints, the constraints are not mandatory for the default resolution policy to work.
> The `DefaultResolutionSet` is the set of resolutions which must be defined in order to create the default resolution constraints for a trackable. When provided, they are used by the default resolution policy to choose a better resolution for a given situation.

**4.1 Default resolution constraints**
- `DefaultResolutionSet` - description does not explain its purpose. Since we're discussing the default implementation we can be more specific. Something like `The set of resolutions for different proximity and amount of active subscribers values`.
- `DefaultProximity` - description suggests that after reaching proximity the resolution is increased and that's not true. If proximity is reached then a certain resolution from the `DefaultResolutionSet` is used, but it does not necessarily have to be faster than the active one, it depends on the configuration. Therefore, I'd rephrase it to show that the proximity defines when to use which resolution from the `DefaultResolutionSet`.

**5. Resolution policy**
- I'd add info about how to create a custom resolution policy, building blocks and maybe examples based on the default implementation.
- Show code snippet of creating the resolution policy and providing it in the publisher builder.

**5.1 Default resolution policy**
- I'd add [the diagram](https://ably-real-time.slack.com/archives/C01EPJENRD0/p1646998727835569?thread_ts=1646906521.453499&cid=C01EPJENRD0) that shows how exactly it works.
- I'd add a few example scenarios to show how the final resolution is calculated.

Missing information:
- Constant location engine resolution - what is it for and how does it work.

-----

https://ably.com/docs/asset-tracking/using-the-sdks

**6.1 Initializing the publisher**
- Required config - missing: notification provider (Android specific, iOS might be different - need to check).
- Optional config - missing: log handler, raw locations, send resolution, predictions, raw history data callback, constant location engine resolution (Android specific, iOS might be different - need to check).
- Code snippet about resolution constraints seems out of place.
- Change the builder code snippet to show the minimum required config and the full optional config as well.

**6.2 Start tracking**
- Missing information about optional destination for a Trackable.
- Missing info about wrapping the `track()/add()` call in `try/catch` for error handling.
- Missing info about differences between `track()` and `add()`.
- Missing info about the return type of `track()/add()` which is a flow of trackable states.
- Describe what happens when you call `add()` with the same `trackableId` but different destination/constraints.
- Describe what happens when you call `track()` with the same `trackableId` but different destination/constraints.

Missing information:
- Other methods of publisher like: `trackables`, `getTrackableState()`, `routingProfile`.

**7.1 Initializing the Subscriber**
- Optional config - missing: log handler.

**7.2 Subscribe to updates**
- Add info about subscribing to raw locations (for debugging purposes only, not suitable for production).

**7.4 Request a different resolution**
- Add info about subscribing to resolutions flow.

Missing information:
- Info about `nextLocationUpdateIntervals` flow for smooth animations.

-----

**Add a section about the Location Animator**
- What problem does it solve (map agnostic, smooth asset animations out of the box for AAT).
- How it works (intentional delay, speeding up/slowing down the animation).
- What can you listen to (trackable and camera positions).
- What you need to provide for it to work (location update + next location update interval).

**Comment by tbedford** [comment]
@KacperKluka with regards the following:

> 5. Resolution policy

> I'd add info about how to create a custom resolution policy, building blocks and maybe examples based on the default implementation.
Show code snippet of creating the resolution policy and providing it in the publisher builder.

Do you have any raw materials we can use for this? For example, do we have suitable code snippets or examples we could use here? Thanks!

**Comment by KacperKluka** [comment]
@tbedford unfortunately, there are no materials already created for that. However, I could prepare them for you, just let me know what you need :wink:

**Comment by tbedford** [comment]
> @tbedford unfortunately, there are no materials already created for that. However, I could prepare them for you, just let me know what you need 😉

Thanks. Let's break it down and start with "how to create a custom resolution policy". If we could create something on how to do that that would be very helpful. Is this something that would be done in code? Or would they start by deciding on the parameters? Basically anything you could provide on this would be helpful. Thanks again!

**Comment by KacperKluka** [comment]
To create your custom resolution policy you have 2 required classes to implement and a few optional ones. The required ones are:
- `ResolutionPolicy` - responsible for calculating network and GPS resolutions
- `ResolutionPolicy.Factory` - responsible for creating the resolution policy instance

The optional ones are:
- `ResolutionPolicy.Hooks.SubscriberSetListener` - called whenever the amount of subscribers changes
- `ResolutionPolicy.Hooks.TrackableSetListener` - called whenever the amount of trackables changes or the active trackable changes
- `ResolutionPolicy.Methods.ProximityHandler` - called when the proximity is reached or cancelled

From within the `ResolutionPolicy` you have access to:
- `ResolutionPolicy.Methods` - you can use it to perform actions on the publisher (e.g. `refresh()` and notify the publisher about some events (e.g. `setProximityThreshold()`) 
- `ResolutionPolicy.Hooks` - you can use it to attach the aforementioned optional listeners

Additionally, you can create custom `ResolutionConstraints` that can later be provided when creating a `Trackable`. This allows you to add constraints to a trackable that later can be used when resolving a resolution by your custom resolution policy (e.g. you can set a battery threshold and modify the frequency of messages based on the current battery level).

After everything is done you have to provide your custom resolution policy factory to the `Publisher` builder:

```kotlin
Publisher.publishers()
    .resolutionPolicy(CustomResolutionPolicyFactory("optional param"))
    // other required configurations
   .start()
```

---

The simplest custom implementation would be
```kotlin
class CustomResolutionPolicyFactory : ResolutionPolicy.Factory {
    override fun createResolutionPolicy(
        hooks: ResolutionPolicy.Hooks,
        methods: ResolutionPolicy.Methods
    ): ResolutionPolicy {
        return CustomResolutionPolicy(hooks, methods)
    }
}

class CustomResolutionPolicy(
    hooks: ResolutionPolicy.Hooks,
    private val methods: ResolutionPolicy.Methods,
) : ResolutionPolicy {
    override fun resolve(resolutions: Set<Resolution>): Resolution =
        resolveGpsResolution(resolutions)

    override fun resolve(request: TrackableResolutionRequest): Resolution =
        resolveNetworkResolution(request)

    private fun resolveGpsResolution(requests: Set<Resolution>): Resolution =
        TODO()

    private fun resolveNetworkResolution(request: TrackableResolutionRequest): Resolution =
        TODO()
}
```

A version with all optional listeners and a custom parameter would look like this:
```kotlin
class CustomResolutionPolicyFactory(
    private val customParameter: String,
) : ResolutionPolicy.Factory {
    override fun createResolutionPolicy(
        hooks: ResolutionPolicy.Hooks,
        methods: ResolutionPolicy.Methods
    ): ResolutionPolicy {
        return CustomResolutionPolicy(hooks, methods, customParameter)
    }
}

class CustomResolutionPolicy(
    hooks: ResolutionPolicy.Hooks,
    private val methods: ResolutionPolicy.Methods,
    private val customParameter: String,
) : ResolutionPolicy {
    private val proximityHandler = CustomProximityHandler()

    init {
        hooks.trackables(CustomTrackableSetListener())
        hooks.subscribers(CustomSubscriberSetListener())
    }

    override fun resolve(resolutions: Set<Resolution>): Resolution =
        resolveGpsResolution(resolutions)

    override fun resolve(request: TrackableResolutionRequest): Resolution =
        resolveNetworkResolution(request)

    private fun resolveGpsResolution(requests: Set<Resolution>): Resolution =
        TODO()

    private fun resolveNetworkResolution(request: TrackableResolutionRequest): Resolution =
        TODO()

    private inner class CustomSubscriberSetListener : ResolutionPolicy.Hooks.SubscriberSetListener {
        override fun onSubscriberAdded(subscriber: Subscriber) {
            // do something
        }

        override fun onSubscriberRemoved(subscriber: Subscriber) {
            // do something
        }
    }

    private inner class CustomTrackableSetListener : ResolutionPolicy.Hooks.TrackableSetListener {
        override fun onTrackableAdded(trackable: Trackable) {
            // do something
        }

        override fun onTrackableRemoved(trackable: Trackable) {
            // do something
        }

        override fun onActiveTrackableChanged(trackable: Trackable?) {
            // do something
        }
    }

    private inner class CustomProximityHandler : ResolutionPolicy.Methods.ProximityHandler {
        override fun onProximityReached(threshold: Proximity) {
            // do something
        }

        override fun onProximityCancelled() {
            // do something
        }
    }
}
```

---

I know that's a lot but I've tried to keep it as short as possible :sweat_smile: @tbedford please let me know if you need anything else :wink:

**Comment by tbedford** [comment]
OK next up...

> Constant location engine resolution - what is it for and how does it work.

I've not seen anything on this. Do we have any info anywhere? Design docs, Confluence? If there's nothing you can point me at, can you provide some info on this please @KacperKluka? Thanks!

**Comment by KacperKluka** [comment]
You can find some explanation in our API docs [here](https://github.com/ably/ably-asset-tracking-android/blob/main/publishing-sdk/src/main/java/com/ably/tracking/publisher/Publisher.kt#L242-L252) :wink: 

In short, this is a mechanism that overwrites the resolution policy's GPS resolution and always uses a static GPS resolution. This can be helpful if, for example, you want to send messages every 10 seconds but you want to retrieve them from the GPS sensor each second, so the quality of your trip data is high. Let me know if you'll have any questions :grin:

**Comment by tbedford** [comment]
> I'd add a few example scenarios to show how the final resolution is calculated.

Sorry @KacperKluka - could you provide one example please of what you mean here. Thanks.

**Comment by KacperKluka** [comment]
Sure :wink: I wanted to show how the `DefaultResolutionPolicy` works on a few scenarios to make it easier for users to understand it.

Scenario 1 (one trackable, no subscribers)
```
Input:
Trackable A resolution constraint = Balanced, 1000, 1.0

Output:
Trackable A network resolution = Balanced, 1000, 1.0
GPS resolution = Balanced, 1000, 1.0
```

Scenario 2 (one trackable, one subscriber)
```
Input:
Trackable A resolution constraint = Balanced, 1000, 1.0
Trackable A subscriber 1 resolution = Maximum, 50, 2.0

Output:
Trackable A network resolution = Maximum, 50, 1.0
GPS resolution = Maximum, 50, 1.0
```

Scenario 3 (one trackable, two subscribers)
```
Input:
Trackable A resolution constraint = Balanced, 1000, 1.0
Trackable A subscriber 1 resolution = Maximum, 50, 2.0
Trackable A subscriber 2 resolution = Low, 10, 5.0

Output:
Trackable A network resolution = Maximum, 10, 1.0
GPS resolution = Maximum, 10, 1.0
```

Scenario 4 (two trackables, no subscribers)
```
Input:
Trackable A resolution constraint = Balanced, 1000, 1.0
Trackable B resolution constraint = High, 2000, 3.0

Output:
Trackable A network resolution = Balanced, 1000, 1.0
Trackable B network resolution = High, 2000, 3.0
GPS resolution = High, 1000, 1.0
```

Scenario 5 (one trackable without constraints, one subscriber)
```
Input:
Trackable A subscriber 1 resolution = Maximum, 50, 2.0

Output:
Trackable A network resolution = Maximum, 50, 2.0
GPS resolution = Maximum, 50, 2.0
```

Scenario 6 (one trackable without constraints, no subscribers)
```
Input:
Default resolution policy resolution = Balanced, 3000, 10.0

Output:
Trackable A network resolution = Balanced, 3000, 10.0
GPS resolution = Balanced, 3000, 10.0
```

**Comment by tbedford** [comment]
@KacperKluka 

Regarding this:

> 6.1 Initializing the publisher

> Required config - missing: notification provider (Android specific, iOS might be different - need to check).
> Optional config - missing: log handler, raw locations, send resolution, predictions, raw history data callback, constant location engine resolution (Android specific, iOS might be different - need to check).
> Code snippet about resolution constraints seems out of place.
> Change the builder code snippet to show the minimum required config and the full optional config as well.

The current docs are based on the README here: https://github.com/ably/ably-asset-tracking-android#publishing-sdk

Does the README need to be updated? For example, `Required config - missing: notification provider` - it doesn't appear to be shown in the README, as the example there is: 

```
// Initialise and Start the Publisher
val publisher = Publisher.publishers() // get the Publisher builder in default state
    .connection(ConnectionConfiguration(Authentication.basic(CLIENT_ID, ABLY_API_KEY))) // provide Ably configuration with credentials
    .map(MapConfiguration(MAPBOX_ACCESS_TOKEN)) // provide Mapbox configuration with credentials
    .androidContext(this) // provide Android runtime context
    .resolutionPolicy(DefaultResolutionPolicyFactory(defaultResolution, this)) // provide either the default resolution policy factory or your custom implementation
    .profile(RoutingProfile.DRIVING) // provide mode of transportation for better location enhancements
    .start()
```

and this is what we have in the docs:

```
// Initialize and Start the Publisher
  val publisher = Publisher.publishers() // get the Publisher builder in default state
      .connection(ConnectionConfiguration(Authentication.basic(CLIENT_ID, ABLY_API_KEY))) // provide Ably configuration with credentials
      .map(MapConfiguration(MAPBOX_ACCESS_TOKEN)) // provide Mapbox configuration with credentials
      .androidContext(this) // provide Android runtime context
      .resolutionPolicy(DefaultResolutionPolicyFactory(defaultResolution, this)) // provide either the default resolution policy factory or your custom implementation
      .profile(RoutingProfile.DRIVING) // provide mode of transportation for better location enhancements
      .start()
```

So, does the README need to be updated? Did I misunderstand something? Thanks!

**Comment by KacperKluka** [comment]
Good catch, it looks like the README needs to be updated as well :sweat_smile:

**Comment by tbedford** [comment]
@KacperKluka - dare I ask, would you be able to do that at some point please? Thank you!!!

**Comment by KacperKluka** [comment]
Here's a [PR](https://github.com/ably/ably-asset-tracking-android/pull/750) with the README changes :wink: Let me know if you'll need more help with this.

**Comment by tbedford** [comment]
@KacperKluka

> Add info about subscribing to raw locations (for debugging purposes only, not suitable for production).

How would they do that? Is this done via code? A debug flag? Would you be able to provide an example of this please? Thanks.

**Comment by KacperKluka** [comment]
It needs to be done in both Publisher and Subscriber SDKs:

**Publisher**
You have to enable the raw locations by calling the `rawLocations(true)` builder method. It will enable sending raw locations from the publisher.

**Subscriber**
Once the raw locations are enabled by the publisher, you can access them via the subscriber's `rawLocations` flow, just as you do with the regular `locations` flow.
